### ISSUE TITLE
Add support for AAC-LC audio format in SDL mixer

### ISSUE THREAD

**Issue by MyreMylar** [libsdl-org/SDL_mixer#338]
We had a user on the pygame github who was unable to play an AAC file using pygame - [issue here](https://github.com/pygame/pygame/issues/2785).

I investigated it a little bit and it looks like, [relatively recently, ffmpeg gained better support](https://trac.ffmpeg.org/wiki/Encode/AAC) for this file format. 

**Why add it?**
 I'm no audio expert but apparently AAC has been Apple's preferred audio codec for quite some time now.

**Why not add it?**

More work and SDL mixer has not had many updates recently anyway. [ There is a better AAC encoder/decoder that has licensing issues](https://superuser.com/questions/1610747/is-fdk-aac-still-better-than-ffmpegs-native-aac-encoder). Lots of alternative audio formats.

I know very little about how the audio stuff all works so no doubt I've missed something obvious; perhaps ffmpeg is not even used for audio playback via SDL mixer. Perhaps AAC-LC actually is supported and there was just something off with this particular file.

Anyway, I thought I'd raise it in case it was of interest for anyone searching in the future.

**Comment by sezero** [comment]
Moved to SDL_mixer issues.

**Comment by slouken** [comment]
Here is the open source AAC decoder library:
https://netactuate.dl.sourceforge.net/project/opencore-amr/fdk-aac/fdk-aac-2.0.2.tar.gz

**Comment by slouken** [comment]
Adding ffmpeg support might be interesting as well.

**Comment by Wohlstand** [comment]
Speaking about FDK-AAC, are you sure it's license is fine?
https://github.com/mstorsjo/fdk-aac/blob/master/NOTICE

**Comment by Wohlstand** [comment]
> Adding ffmpeg support might be interesting as well.

Recently, at Discord chat, we talked with friends against FFMPEG support adding (mainly for WMA support adding which is a controversial issue, because there are a very small amount of SMBX episodes that uses WMA files, almost nobody uses them), however, FFmpeg itself is a big and heavy thing for static linking (if not try to build the minimalistic version with everything disabled except specific file formats). But could be supported easily if just use from the system. Anyway, FFMPEG using on Windows is a pain, because it's hard to build it as needed to provide a ton of important external dependencies. I do have experience with FFmpeg usage because of my main job, I could try to implement some (and then, you could re-use that) at one random moment when I have a free time.

**Comment by slouken** [comment]
> Speaking about FDK-AAC, are you sure it's license is fine? https://github.com/mstorsjo/fdk-aac/blob/master/NOTICE

Yes. SDL can actually use any licensed code, but by default we try to include code that doesn't impose any restrictions on the license of the including code. So MIT/BSD/Zlib, etc. are all fine to enable by default. LGPL/GPL, etc. are allowed to be included, but must not be compiled in by default.

**Comment by Wohlstand** [comment]
From my side: not yet. Maybe I'll try to add the just dynamical use of existing in-system FFMPEG thing. Or just both FFMPEG and AAC-LC for user choice. I do take into MixerX these libraries I am able to bundle into AudioCodecs thing to simplify build of the whole chain if dependencies at the same place, and I rarely take libraries that are heavy or hard to include into AudioCodecs due to lack of CMake build vendored

**Comment by Wohlstand** [comment]
Hi there!
At my side of MixerX fork I started an experiment to add the FFMPEG support (incomplete yet, some polishing and clean-up are needed): https://github.com/WohlSoft/SDL-Mixer-X/tree/wip-ffmpeg-support
And it works for AAC, MP4 files, Opus files inside WebM container, and WMA files. However, there is a problem with seeking WMA and WebM-Opus files, and I don't know the reason for failure yet (isn't it a bug of FFMPEG itself?). I also had to run some audio conversions using libswresample to merge planar audio (which happens on WMA and some AAC files) into a normal stream acceptable by SDL Audio.

The code itself is here: https://github.com/WohlSoft/SDL-Mixer-X/blob/wip-ffmpeg-support/src/codecs/music_ffmpeg.c

**Comment by icculus** [comment]
Some notes, since I'm going through all the SDL_mixer bug reports:

- fdk-aac is probably fine licensing-wise. It has that obnoxious "you have to mention us in your documentation" clause that drives me nuts, but that doesn't stop SDL_mixer from implementing a decoder that uses it. We just simply don't ship the external library, and if the person using SDL_mixer doesn't ship it either, they get no AAC support but have no licensing requirements. If the library is already installed on the system (like a Linux distro package), that's someone else's licensing problem.  :)
- AAC has patents and Fraunhofer theoretically enforces them. However, _most of the patents are expired_. Like most of these situations, there are 14 million little patents that extend past the original, and you need a few years for those to expire too for complete coverage, and the details are often not entirely clear. One google search seems to think that most of what you care about in AAC is already expired, and the last bits of some streaming stuff expires around 2028. Again, even if none of them have expired, it's not a reason to not write a decoder, as the same rules as the software license apply.
- Someone should write a public domain single-header library for AAC.
- ffmpeg and faad2 are GPL'd code, they are non-starters for me.

**Comment by Wohlstand** [comment]
Continuing my past post (after more than 3 years!), I finished ffmpeg support a while ago, and it's just a native part of the MixerX for a while, however, I don't yet distribute builds since I need to finish a "frog task" to implement CMake wrapper over FFMPEG build to ensure it builds on every supported platform (not just desktops, but also several bare metal targets like 3DS, Wii, Wii U, Switch, Vita, and Android). The biggest pity it doesn't offers own CMake build, and this adds serious dependency on GNU Make that is a pain at platforms that doesn't offer it by default (MSVC on Windows, BSD-based systems, etc.).

> ffmpeg and faad2 are GPL'd code, they are non-starters for me.

I smoothly remember that AAC support is enough with LGPL mode, but, I going to verify to be fully sure...

**Comment by Wohlstand** [comment]
Okay, according to heads, AAC support is enough with LGPL mode:
- https://git.ffmpeg.org/gitweb/ffmpeg.git/blob/HEAD:/libavcodec/aac/aacdec.c
- https://git.ffmpeg.org/gitweb/ffmpeg.git/blob/HEAD:/libavcodec/aac/aacdec_ac.c
- https://git.ffmpeg.org/gitweb/ffmpeg.git/blob/HEAD:/libavcodec/aac/aacdec_fixed.c
- And just entire directory: https://git.ffmpeg.org/gitweb/ffmpeg.git/tree/HEAD:/libavcodec/aac

The license of ffmpeg depends on which codecs you will enable. It even has the "non-free" mode that allows to enable some modules that usually never being built, and it even warns that build is "non-redistributable" because of licenses conflict in the done build.

**Comment by Wohlstand** [comment]
> patents

I know that software patents are means nothing at Europe and lots of other countries. At European Union, as I know, it's very conditionally possible to make a patent for software (In condition if it makes easier work with PC like improvement of memory access mechanism, etc. and has the "innovative step"). At my country software only can be registered as a copyright property, but no patents for software algorithms. However, if distribute stuff for American market, this makes a serious pain and need to make sure they are cared properly.

Also: FFMPEG is made by Europeans, and therefore it lives very long without any attacks by patents.

**Comment by icculus** [comment]
(Taking this out of 3.2.0 milestone: would be nice to have, but isn't crucial.)

**Comment by mjdave** [comment]
Hi, I'm just now making a music player using SDL3, using SDL_Mixer from the main branch.

For me, I need AAC support one way or another. It wasn't clear that SDL_Mixer wouldn't play .m4a files, and then not clear why not, until I found this thread here just now.

So now I better understand the issues and the complexities, I would just like to still put in my vote for some AAC support to be added, and also some documentation listing the supported formats somewhere like [here](https://wiki.libsdl.org/SDL3_mixer/CategorySDLMixer) would be great.

I'm also still looking for a solution for AAC playback, and am keen to help if I can.
